openshift/WFLY-15421_OIDC_for_WildFly_s2i_v2.adoc

```diff
@@ -0,0 +1,297 @@
+
+* https://github.com/wildfly/wildfly-s2i[WildFly s2i]
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [ ] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [ ] Managed domain
+
+* [x] OpenShift s2i
+
+* [ ] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+==== Galleon layer for oidc support
+
+This proposal relies on the WildFly `elytron-oidc-client` Galleon layer. This layer must be present when provisioning a server 
+in order for this feature to be available.
+
+==== Descriptor based configuration
+
+This support works out of the box and doesn’t require any addition to the WildFly s2i image. It applies to any kind of OIDC provider.
+
+Users can define expressions in the json file in order to inject values at runtime.
+For example, `provider-url: ${env.OIDC_URL}` would imply that the env variable `OIDC_URL=http://keycloak/auth/realms/WildFly` is set.
+
+NB: This way of configuring a deployment implies that a “client” has been created in the authorization server. That is expected.
+
+==== User CLI scripts called at packaging time
+
+When using the https://issues.redhat.com/browse/WFLY-14934[WildFly Maven plugin] to package an application, a script can be called to configure the OIDC subsystem for the deployed application.
+
+This support works out of the box and doesn’t require any addition to the WildFly s2i image. It applies to any kind of OIDC provider.
+
+This implies a set of WildFly CLI management operations to be called at packaging time.
+
+For example:
+`/subsystem=elytron-oidc-client/secure-deployment=simple-webapp.war:add(client-id=simple-webapp, public-client=true, provider-url=${OIDC_URL}, ssl-required=EXTERNAL)`
+
+NB: This way of configuring a deployment implies that a “client” has been created in the authorization server. That is expected.
+
+==== Environment variable based configuration
+
+This is the main topic of this proposal.
+
+A CLI script is generated at boot time to handle the OIDC subsystem configuration based on discovered 
+deployments (actually deployed and found in content repository or located in `standalone/deployments` directory).
+In addition there is an automatic creation and registration of the clients in the authorization server.
+
+In order for war deployments to be discovered they must have the `OIDC` auth-method in web.xml. 
+Deployments that contain a `WEB-INF/oidc.json` descriptor file are ignored.
+
+NB: A single provider can be configured for a running server. The discovered deployments are all registered inside the same OIDC provider.
+
+We are offering a new set of env variables `OIDC_*` to be used to configure the subsystem, enable the discovery of deployments and registration of 
+clients into the authorization server.
+
+The existing `SSO_*` env variables used for OIDC are usable for Keycloak OIDC integration. 
```

Comment:
We should also deprecate all SSO_* env variables, wdyt?

Reply:
Correct, fixed.